Move requires_grad_/backward out of VariableTypeManual.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54543 Move requires_grad_/backward out of VariableTypeManual.**
* #54459 Move some variable ops out of VariableTypeManual.

Differential Revision: [D27321819](https://our.internmc.facebook.com/intern/diff/D27321819)

Commit:
Move requires_grad_ and backward off dispatcher.

[ghstack-poisoned]

Commit:
Update on "Move requires_grad_/backward out of VariableTypeManual."




[ghstack-poisoned]

Commit:
Update on "Move requires_grad_/backward out of VariableTypeManual."




[ghstack-poisoned]

ezyang commented on test/cpp/api/tensor.cpp:
```diff
@@ -1046,7 +1046,7 @@ TEST(TensorTest, RequiresGradInplace) {
   ASSERT_FALSE(x.requires_grad());
 
-  const auto int_tensor = torch::tensor({5}, at::TensorOptions().dtype(torch::kInt));
+  auto int_tensor = torch::tensor({5}, at::TensorOptions().dtype(torch::kInt));
```
hmm?

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -81,7 +81,7 @@
 
 - func: requires_grad_(Tensor(a!) self, bool requires_grad=True) -> Tensor(a!)
-  manual_kernel_registration: True
+  manual_cpp_binding: True
```
Why does this need manual binding?

ailzhang replied:
I put them in manual binding mainly because these functions are only manipulating autograd related attributes on tensor instead of doing real operation with the tensor data. So I feel we could take some perf gain by moving them to manual binding :D

Commit:
Update on "Move requires_grad_/backward out of VariableTypeManual."




[ghstack-poisoned]

Commit:
Update on "Move requires_grad_/backward out of VariableTypeManual."


Differential Revision: [D27321819](https://our.internmc.facebook.com/intern/diff/D27321819)

[ghstack-poisoned]

Commit:
Update on "Move requires_grad_/backward out of VariableTypeManual."


Differential Revision: [D27321819](https://our.internmc.facebook.com/intern/diff/D27321819)

[ghstack-poisoned]

Commit:
Update on "Move requires_grad_/backward out of VariableTypeManual."


Differential Revision: [D27321819](https://our.internmc.facebook.com/intern/diff/D27321819)

[ghstack-poisoned]
